### DISCUSSION TITLE
Docker Run Images with option via Golang Docker SDK

### DISCUSSION THREAD

**Discussion by DineshGuptaa** [moby/moby#43964] [Q&A] [answered]
Hi All,
I am using docker  and wanted to run it pragmatically. I have an image taken vnc images from [here](https://github.com/ConSol/docker-headless-vnc-container) and trying to run it from Golang program.
When run docker from Ubuntu CLI then I am able to see the output
http://192.168.53.120/ and able to see the vnc desktop.  Command I am running as 
docker run -d -p 5901:5901 -p 6901:6901 -p 8888:8888/udp -u 0 ubuntu-xfce-vnc 
But same thing I am unable to achieve via Golang and Docker API

    package main
    import (
        "fmt"
        "github.com/docker/docker/api/types/container"
        "github.com/docker/docker/api/types"
        "github.com/docker/docker/api/types/filters"
        "github.com/docker/docker/client"
        "golang.org/x/net/context"
     )

     func main() {
        cli, err := client.NewClientWithOpts(client.WithVersion("1.38"))
        if err != nil {
                panic(err)
        }
        ctx := context.Background()
        args := filters.NewArgs(filters.KeyValuePair{"reference","gavnc:latest"})
        imgs, err := cli.ImageList(ctx, types.ImageListOptions{Filters: args})

        if err != nil {
                panic(err)
        }
        //fmt.Println("RepoTags \t Created \t Size")
        for _, img := range imgs {
                fmt.Println(img.RepoTags[0], img.Created, img.Size)
                imgTag := img.RepoTags[0]
        if imgTag != ""{
        println("Image Found....Now Running Image...")
            resp, err := cli.ContainerCreate(ctx, &container.Config{
                Image: imgTag,
            }, nil, nil, "")
        if err != nil {
            panic(err)
        }
        println("Id of image:",resp.ID)
        err = cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{});
        if err != nil {
            panic(err)
        }
            }
        }
     }

Please guide me how we may pass the option parameters. I am unable to find sample for the same.

Thansk

**Comment by hardiksakhuja** [comment]
Wow ,  3 years and no answers . 
I faced a similar challenge while performing docker prune . 
and after much digging , this worked . 

`	pruneFilters := filters.NewArgs()
	pruneFilters.Add("dangling", "false")`
	
you might want to take a look at 
[https://golang.hotexamples.com/examples/github.com.docker.docker.api.types.filters/-/NewArgs/golang-newargs-function-examples.html](url)
to get more ideas .

**Comment by iamjoemccormick** [comment]
For anyone else that lands here from Google trying to figure out how to filter images using ImageListOptions. The filters are based on what is available in the Docker API for [ImageList](https://docs.docker.com/engine/api/v1.41/#tag/Image/operation/ImageList), not what is available on types.ImageSummary or anything else in the Go SDK. 

For example if you wanted to find a particular image this won't work;

```go
f := filters.NewArgs()
f.Add("RepoTags", "debian:10.12")
```
You have to do:
```go
f.Add("reference", "debian:10.12")
```
Then you can get the results like this: 
```go
images, _ := m.client.ImageList(m.ctx, types.ImageListOptions{All: false, Filters: f})
```

**Comment by cpuguy83** [comment, marked as answer]
The way `docker run` works is it first calls the create endpoint, if the image isn't available then it returns a 404.
`docker run` sees that 404 and initiates a pull, then does another create.
